hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java

```diff
@@ -0,0 +1,370 @@
+
+    // There should be no residual since we already negotiated that earlier in
+    // HBaseStorageHandler.decomposePredicate. However, with hive.optimize.index.filter
+    // OpProcFactory#pushFilterToStorageHandler pushes the original filter back down again.
+    // Since pushed-down filters are not omitted at the higher levels (and thus the
+    // contract of negotiation is ignored anyway), just ignore the residuals.
+    // Re-assess this when negotiation is honored and the duplicate evaluation is removed.
+    // THIS IGNORES RESIDUAL PARSING FROM HBaseStorageHandler#decomposePredicate
+    if (residualPredicate != null) {
+      LOG.debug("Ignoring residual predicate " + residualPredicate.getExprString());
+    }
+
+    Map<String, List<IndexSearchCondition>> split = HiveHBaseInputFormatUtil.decompose(conditions);
+    List<IndexSearchCondition> keyConditions = split.get(keyColName);
+    if (keyConditions != null && !keyConditions.isEmpty()) {
+      HiveHBaseInputFormatUtil.setupKeyRange(scan, keyConditions, isKeyBinary);
+    }
+    List<IndexSearchCondition> tsConditions = split.get(tsColName);
+    if (tsConditions != null && !tsConditions.isEmpty()) {
+      HiveHBaseInputFormatUtil.setupTimeRange(scan, tsConditions);
+    }
+    return scan;
+  }
+
+  /**
+   * Instantiates a new predicate analyzer suitable for
+   * determining how to push a filter down into the HBase scan,
+   * based on the rules for what kinds of pushdown we currently support.
+   *
+   * @param keyColumnName name of the Hive column mapped to the HBase row key
+   *
+   * @return preconfigured predicate analyzer
+   */
+  static IndexPredicateAnalyzer newIndexPredicateAnalyzer(String keyColumnName, boolean isKeyComparable,
+      String timestampColumn) {
+
+    IndexPredicateAnalyzer analyzer = new IndexPredicateAnalyzer();
+
+    // We can always do equality predicate. Just need to make sure we get appropriate
+    // BA representation of constant of filter condition.
+    // We can do other comparisons only if storage format in hbase is either binary
+    // or we are dealing with string types since there lexicographic ordering will suffice.
+    if (isKeyComparable) {
+      analyzer.addComparisonOp(keyColumnName, "org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPEqual",
+          "org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPEqualOrGreaterThan",
+          "org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPEqualOrLessThan",
+          "org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPLessThan",
+          "org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPGreaterThan");
+    } else {
+      analyzer.addComparisonOp(keyColumnName, "org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPEqual");
+    }
+
+    if (timestampColumn != null) {
+      analyzer.addComparisonOp(timestampColumn, "org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPEqual",
+          "org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPEqualOrGreaterThan",
```

Comment:
Maybe wrong question but it is not a good idea to validate if this UDF's are presented? Maybe it is too defensive.

Reply:
For this, actually the function is from https://github.com/apache/hive/blob/master/hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormat.java#L238. I would not change it in this PR. For me, I am not sure if its easy to detect the UDF, and maybe we can use a new thread to track if you think its a good optimization?